acl/acl.go

```diff
@@ -54,141 +24,61 @@ const (
 	ActionDelete    = Action("delete")
 )
 
-// Policy is an ACL policy that controls access to secrets.
-type Policy struct {
-	rules []compiledRule
-}
-
-// Compile returns a Policy that enforces the ACLs in bs.
-func Compile(bs []byte) (*Policy, error) {
-	bs, err := hujson.Standardize(bs)
-	if err != nil {
-		return nil, fmt.Errorf("converting ACL policy to JSON: %w", err)
-	}
-
-	type rule struct {
-		From   []string `json:"principal"`
-		Secret []string `json:"secret"`
-		Action []Action `json:"action"`
-	}
-	var pol struct {
-		Rules []rule `json:"rules"`
-	}
-	dec := json.NewDecoder(bytes.NewBuffer(bs))
-	dec.DisallowUnknownFields()
-	if err := dec.Decode(&pol); err != nil {
-		return nil, fmt.Errorf("parsing ACL policy JSON: %w", err)
-	}
-
-	ret := &Policy{}
-	var errs []error
-
-	for i, r := range pol.Rules {
-		ruleNum := i + 1
-
-		from, err := expandFrom(ruleNum, r.From)
-		if err != nil {
-			errs = append(errs, err)
-		}
-		secret, err := expandSecret(ruleNum, r.Secret)
-		if err != nil {
-			errs = append(errs, err)
-		}
-		action, err := expandAction(ruleNum, r.Action)
-		if err != nil {
-			errs = append(errs, err)
-		}
-		ret.rules = append(ret.rules, compiledRule{from, secret, action})
-	}
-
-	if len(errs) > 0 {
-		return nil, multierr.New(errs...)
-	}
-	return ret, nil
-}
+// Rules is a set of ACLs for access to a secret.
+type Rules []Rule
 
-// Allow reports whether anyone in 'from' can perform 'action' on 'secret'.
-func (p *Policy) Allow(from []string, secret string, action Action) bool {
-	for _, r := range p.rules {
-		if r.allow(from, secret, action) {
+// Allow reports whether the ACLs allow action on secret.
+func (rr Rules) Allow(action Action, secret string) bool {
+	for _, r := range rr {
+		if r.Allow(action, secret) {
 			return true
 		}
 	}
 	return false
 }
 
-type compiledRule struct {
-	from   map[string]bool
-	secret *regexp.Regexp
-	action map[Action]bool
+// Rule is an access control rule that permits some actions on some
+// secrets. Secrets can contain '*' wildcards, which match zero or
+// more characters.
+type Rule struct {
+	Action []Action `json:"action"`
+	Secret []string `json:"secret"`
```

Comment:
Something to think about for a future change, we could define a little wrapper `type Secret string` that has a `Match` method to do the glob matching.

Reply:
Good call, I'll make a quick followup.